### ISSUE TITLE
`timeToLiveSeconds` not being respected

### ISSUE THREAD

**Issue by aodj** [kubeflow/spark-operator#1761]
We've been using the spark-operator for some time and often see FAILED jobs that are not deleted by the operator. The oldest I have seen is a FAILED job that's 300+ days old. The operator is configured with the default Helm chart value of `resyncInterval: 30` so I would expect to see it garbage collect these pods.

We do see the occasional cleanup of SparkApplication jobs, but often not.

**Comment by julienlau** [comment]
Hi,

You mentionned timeToLiveSeconds in the title, but from your message it seems you don't always put it in the sparkApp manifest.
Could you please clarify ?

What about the manifest of the failed jobs from > 300d old ?
Does it contains :

```
spec:
  timeToLiveSeconds: xxx
```

?

**Comment by aodj** [comment]
Each of the FAILED applications I can check on have `timeToLiveSeconds` set to 3600 and are more than a month old.

```
$ k get sparkapplication -o json | jq '.items[] | .spec.timeToLiveSeconds'
3600
3600
3600
$ k get sparkapplication
NAME                                STATUS   ATTEMPTS   START                  FINISH       AGE
spark-features-2023-03-21-1501-23   FAILED              2023-05-23T12:50:19Z   <no value>   37d
spark-features-2023-03-21-1524-33   FAILED              2023-05-24T10:51:49Z   <no value>   37d
spark-features-2023-03-21-1526-34   FAILED              2023-05-24T13:49:48Z   <no value>   36d
```

**Comment by jacobsalway** [comment]
@Cian911 would you be able to post the `SparkApplication` spec of the apps you're having this issue on? And any more detail as relevant so we can attempt to reproduce.

**Comment by ChenYi015** [comment]
> I have not tested v2 yet, but could this also occur here?

@Cian911 The v2 version also has the same problem, I will raise a PR to fix it.

**Comment by Cian911** [comment]
> > I have not tested v2 yet, but could this also occur here?
> 
> @Cian911 The v2 version also has the same problem, I had raised a PR to fix it.
> 
> Also, could you describe the whole lifecycle of your application that did not get cleaned up? I have tested that the TTL works as expected under the following cases:
> 
> * new -> submit failed -> failed
> * new -> running -> failing -> failed
> * new -> running -> succeeding -> succeeded

Sorry for the delay in replying. In my case, I see this issue under the following cases - also please bare in mind I am using version `spark-operator-chart-1.4.6`.

> new -> submit failed -> failed

Example status output:

```
Status:
  Application State:
    Error Message:  failed to run spark-submit for SparkApplication <redacted>: 24/09/18 17:56:11 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
24/09/18 17:56:12 INFO SparkKubernetesClientFactory: Auto-configuring K8S client using current context from users K8S config file
24/09/18 17:56:16 INFO KerberosConfDriverFeatureStep: You have not specified a krb5.conf file locally or via a ConfigMap. Make sure that you have the krb5.conf locally on the driver image.

    State:  FAILED
  Driver Info:
  Last Submission Attempt Time:  2024-09-18T17:56:19Z
  Submission Attempts:           1
  Termination Time:              <nil>
Events:                          <none>
```

As above, there's no termination time set, which is why the TTL never cleans up the failed submission. I have also seen the same behaviour  in the `new -> running -> failing -> failed` case also, but I do not have an example to hand.

In any case, I think your PR should rectify.

@ChenYi015 Do you have an idea as to why v2 might move out of RC and released officially?

**Comment by ChenYi015** [comment]
@Cian911 Thanks for the detailed information. This issue will be fixed by #2165.

> Do you have an idea as to why v2 might move out of RC and released officially?

 We are going to release the 2.0.0 version this week, after this issue has been fixed.
